### PR TITLE
[Handshake] Add script for visually debugging a handshake program

### PR DESCRIPTION
**Pull request by mortbopet**
This is a script which couples together a .dot representation of a CIRCT
handshake program, and a .vcd trace of a simulation on the same program.
This is done by correlating the .vcd hierarchy with the DOT hierarchy, and
correlating the VCD signals with what looks to be the same values in the DOT
file. Since .dot vertices are created on the handshake operation level,
the script relies on additional attributes in the .dot file to disambiguated
which in- and output edges of a node corresponds to which signal in the .vcd.

.dot files are rendered to .svg images and served by a web server. This server
runs a simple web page which will automatically update the shown image.

The python console provides a simple interface; left and right changes .vcd time +/-1, and 'g' is 'goto ${cycle}'. 
**Disclaimer:** I am **not** a web programmer; just threw something together which would allow for continuous updating of the underlying image + display a cycle count.

Example 1: the following function, HLS'ed through handshake generates ~1800 signals in the .vcd trace:
```mlir
func @exponent(%arg0: i32, %arg1: i32) -> (i32) {
  %c-1_i32 = constant -1 : i32
  %c1_i32 = constant 1 : i32
  %0:2 = scf.while (%arg2 = %arg0, %arg3 = %arg1) : (i32, i32) -> (i32, i32) {
    %1 = cmpi sgt, %arg3, %c1_i32 : i32
    scf.condition(%1) %arg2, %arg3 : i32, i32
  } do {
  ^bb0(%arg2: i32, %arg3: i32):  // no predecessors
    %1 = muli %arg2, %arg0 : i32
    %2 = addi %arg3, %c-1_i32 : i32
    scf.yield %1, %2 : i32, i32
  }
  return %0#0 : i32
}
```
And has the following visual representation:
exponent

Example 2: Also supports hierarchy (there's still a few number of connections which don't get properly indexed but that is a WIP):
instance

### CODE REVIEW

**Comment by mikeurbach** [comment]
This is super cool. I need to look more on the implementation but in general I am super interested in tying together simulation and visualization.

**Comment by teqdruid** [comment]
This is very super cool! I'm looking to do something similar for ESI. Could also be used more generically, with customizations for each dialect. Again, super cool!

**Comment by mortbopet** [comment]
> I'm looking to do something similar for ESI. Could also be used more generically, with customizations for each dialect.

Things would probably translate very easily to ESI. Again, this is just a draft, but i could try and expand it a bit to factor out "vcd trace to dot edge" resolution.

**Comment by teqdruid** [comment]
We can work towards something more general once it's merged. Exciting!

**Comment by mortbopet** [comment]
> We can work towards something more general once it's merged. Exciting!

Sounds good!

I've factored out some classes into separate files to separate concerns a little bit. The PR has now been marked for review @mikeurbach @teqdruid.

File: tools/hsdbg/hsdbg.py [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+            that quite a few assumptions are placed on the format of both the vcd
+            file and the dot file.
+        """
+    # Step 1: infer top module name
+    top = "TOP"
+
+    hier = resolveVCDHierarchy(self.vcd)
+    if len(hier) != 1 or "TOP" not in hier:
+      raise Exception("Expected \"TOP\" as top-level in vcd hierarchy")
+    topHier = hier["TOP"]
```
Comment (mikeurbach) [inline tools/hsdbg/hsdbg.py:146]: I think you can relax this to just enforce there is only one top-level, and take that, similar to how you find dutHier below. I generated a VCD with a different top name, and this works:

```suggestion
    if len(hier) != 1:
      raise Exception("Expected one top-level in vcd hierarchy")
    top = list(hier.keys())[0]
    topHier = hier[top]
```

File: tools/hsdbg/hsdbg.py [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+    self.updateCurrentCycle(cycle)
+
+
+def start_interactive_mode(port, hsdbg):
+  msg = f"""=== Handshake interactive simulation viewer ===
+
+Usage:
+    Open "localhost:{port}" in a browser to view the simulation.
+
+    left arrow: step forward 1 step in vcd time
```
Comment (mikeurbach) [inline tools/hsdbg/hsdbg.py:259]: I think you have the docstring reversed for left versus right here.

**Comment by mikeurbach** [comment]
This is super cool. Left some suggestions above about usability. With those I'm able to load this for a VCD I produced from Vivado for this test: https://github.com/llvm/circt/blob/main/integration_test/Dialect/Handshake/dot.mlir.

However, the UI becomes unresponsive if I try to say `g 1000`. I see some printouts on the JavaScript console, so something may be in a loop? I watched for about 30s but the UI never caught up. I'm attaching the dot and VCD if you're interested (had to add .txt for GH to accept them).

In general, I am super interested in this kind of work. Down the road, I'd love to see the LLHD simulator integrated into a visualization like this in a deep (i.e. embed a library) way. This is a super great start down that path! For a testing tool like this, I don't have any major feedback on the code. I have wanted this while debugging Handshake deadlocks!

[dump.vcd.txt](https://github.com/llvm/circt/files/7565188/dump.vcd.txt)
[top.dot.txt](https://github.com/llvm/circt/files/7565189/top.dot.txt)

**Comment by mortbopet** [comment]
> However, the UI becomes unresponsive if I try to say g 1000

I'm getting the same, `input(...)` seems to be blocking for some reason - although only through VSCode. Running through the terminal, everything seems to work.

**Comment by mortbopet** [comment]
Added a script to drive it from the build directory, as well as some sanity-check integration tests. The tool is disabled by default due to the python package dependencies. The failing CI check now is due to the python linter not liking the change in the `.in` file.
From my end, i think this is good to go as a first take on this tool, so let me know if you have any further comments and/or you are okay with a merge.

**Comment by mikeurbach** [comment]
I still can't get this to work for me, it is the same behavior I described above. This is running the tool in a terminal and viewing the UI in a browser.

Also, pretty minor, but I noticed that asking the tool to quit doesn't actually quit. It prints `Exiting...` but hangs in `lock.acquire()` in `threading.py` until I hit Control-C.

Also, I just noticed the camel case in the Python. In CIRCT we are committing to the Google style: https://github.com/llvm/circt/blob/main/.style.yapf. This prefers snake case for methods and variables.

Please make sure the Python style check passes if you plan to check this in. You can add something to .yapfignore if it is checking files that aren't actual python sources.

**Comment by mortbopet** [comment]
Closing for now; moved this to circt-hls for further incubation.

**Pull request closed without merging**